client/iotcore/client.go

```diff
@@ -307,6 +314,27 @@ func (c *client) DeleteDevice(
 		}
 	}
 
+	if err == nil && len(respListThingPrincipals.Principals) > 0 {
+		// wait for DetachThingPrincipal operation to complete
+		// this operation is asynchronous, so wait couple of seconds
+		for retries := 0; retries <= detachThingPrincipalWaitMaxRetries; retries++ {
+			respListThingPrincipals, err = svc.ListThingPrincipals(ctx,
+				&iot.ListThingPrincipalsInput{
+					ThingName: aws.String(deviceID),
+				})
+			if err != nil {
+				break
+			}
+			if len(respListThingPrincipals.Principals) > 0 {
+				time.Sleep(detachThingPrincipalWaitSleep)
+			}
+		}
+		// Thing Principle still not detached; return error
+		if len(respListThingPrincipals.Principals) > 0 {
+			return ErrThingPrincipalNotDetached
+		}
```

Comment:
This will make the endpoint DELETE endpoint extremely slow in the UI. Perhaps we can improve this later by handling it asynchronously.